src/Arcus.WebApi.Logging/Correlation/CorrelationMiddleware.cs

```diff
@@ -50,94 +40,15 @@ public async Task Invoke(HttpContext httpContext)
             Guard.For<ArgumentException>(() => httpContext.Response is null, "Requires a 'Response'");
             Guard.For<ArgumentException>(() => httpContext.Response.Headers is null, "Requires a 'Response' object with headers");
 
-            if (httpContext.Request.Headers.TryGetValue(_options.Transaction.HeaderName, out StringValues transactionIds))
+            if (_service.TryHttpCorrelate(out string errorMessage))
             {
-                if (!_options.Transaction.AllowInRequest)
-                {
-                    _logger.LogError("No correlation request header '{HeaderName}' for transaction ID was allowed in request", _options.Transaction.HeaderName);
-                    httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
-                    await httpContext.Response.WriteAsync($"No correlation transaction ID request header '{_options.Transaction.HeaderName}' was allowed in the request");
-
-                    return;
-                }
-
-                _logger.LogTrace("Correlation request header '{HeaderName}' found with transaction ID '{TransactionId}'", _options.Transaction.HeaderName, transactionIds);
+                await _next(httpContext);
             }
-
-            string operationId = DetermineOperationId(httpContext);
-            string transactionId = DetermineTransactionId(transactionIds);
-            var correlation = new CorrelationInfo(operationId, transactionId);
-            httpContext.Features.Set(correlation);
-
-            AddCorrelationResponseHeaders(httpContext, operationId, transactionId);
-
-            await _next(httpContext);
-        }
-
-        private string DetermineOperationId(HttpContext httpContext)
-        {
-            if (String.IsNullOrWhiteSpace(httpContext.TraceIdentifier))
+            else
             {
-                string operationId = _options.Operation.GenerateId();
-                if (String.IsNullOrWhiteSpace(operationId))
-                {
-                    throw new InvalidOperationException(
-                        $"Correlation cannot use '{nameof(_options.Operation.GenerateId)}' to generate an operation ID because the resulting ID value is blank");
-                }
-
-                return operationId;
+                httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
```

Comment:
This is a behavioral change which should not be the case I think. Is there a reason why this was changed? When our correlation fails I don't expect it to give a bad request back?

Reply:
Ah, no, it's not actually a behavior change. It's something that's being moved. It was first being set direclty in the middleware but since we now have extracted the common logic, we haven't got access to certain models. (like in the azure functions you should create an `IActionResult` as respnose but not in the middleware).